File: app/src/main/java/morawski/rafal/barcodescanner/scanner/BarcodeBoxDetector.java
package morawski.rafal.barcodescanner.scanner;

import android.content.res.Configuration;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.Rect;
import android.graphics.YuvImage;
import android.util.SparseArray;

import com.google.android.gms.common.images.Size;
import com.google.android.gms.vision.Detector;
import com.google.android.gms.vision.Frame;
import com.google.android.gms.vision.barcode.Barcode;

import java.io.ByteArrayOutputStream;

public class BarcodeBoxDetector extends Detector<Barcode> {
    private static final String TAG = "BarcodeBoxDetector";

    private Detector<Barcode> mDelegate;
    private ScannerBoxOverlay mScannerBoxOverlay;

    public BarcodeBoxDetector(Detector<Barcode> delegate, ScannerBoxOverlay scannerBoxOverlay) {
        mDelegate = delegate;
        mScannerBoxOverlay = scannerBoxOverlay;
    }

    @Override
    public SparseArray<Barcode> detect(Frame frame) {
        if (mScannerBoxOverlay == null) {
            return mDelegate.detect(frame);
        }

        int frameWidth = frame.getMetadata().getWidth();
        int frameHeight = frame.getMetadata().getHeight();
        int frameRotation = frame.getMetadata().getRotation();

        int viewWidth = mScannerBoxOverlay.getWidth();
        int viewHeight = mScannerBoxOverlay.getHeight();

        float mWidthScaleFactor = 1.0f;
        float mHeightScaleFactor = 1.0f;
        if (isPortrait()) {
            mWidthScaleFactor = (float) frameHeight / (float) viewWidth;
            mHeightScaleFactor = (float) frameWidth / (float) viewHeight;
        } else {
            mWidthScaleFactor = (float) frameWidth / (float) viewWidth;
            mHeightScaleFactor = (float) frameHeight / (float) viewHeight;
        }

        Size boxSize = mScannerBoxOverlay.getBoxSize();

        int boxWidth = (int) (boxSize.getWidth() * mWidthScaleFactor);
        int boxHeight = (int) (boxSize.getHeight() * mHeightScaleFactor);

        Rect rect = getRectangle(frameWidth, frameHeight, boxWidth, boxHeight);

        YuvImage yuvImage = new YuvImage(frame.getGrayscaleImageData().array(), ImageFormat.NV21, frameWidth, frameHeight, null);
        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        yuvImage.compressToJpeg(rect, 100, outputStream);
        byte[] jpegArray = outputStream.toByteArray();
        Bitmap bitmap = BitmapFactory.decodeByteArray(jpegArray, 0, jpegArray.length);

        Frame croppedFrame = new Frame.Builder()
                .setBitmap(bitmap)
                .setRotation(frameRotation)
                .build();

        return mDelegate.detect(croppedFrame);
    }

    private Rect getRectangle(int frameWidth, int frameHeight, int boxWidth, int boxHeight) {
        int left = 0, top = 0, right = 0, bottom = 0;

        if (isPortrait()) {
            left = (int) (frameWidth / 2.0 - boxHeight / 2.0);
            top = (int) (frameHeight / 2.0 - boxWidth / 2.0);
            right = (int) (frameWidth / 2.0 + boxHeight / 2.0);
            bottom = (int) (frameHeight / 2.0 + boxWidth / 2.0);

        } else {
            left = (int) (frameWidth / 2.0 - boxWidth / 2.0);
            top = (int) (frameHeight / 2.0 - boxHeight / 2.0);
            right = (int) (frameWidth / 2.0 + boxWidth / 2.0);
            bottom = (int) (frameHeight / 2.0 + boxHeight / 2.0);
        }

        return new Rect(left, top, right, bottom);
    }

    private boolean isPortrait() {
        int orientation = Resources.getSystem().getConfiguration().orientation;
        return orientation == Configuration.ORIENTATION_PORTRAIT;
    }
}

File: app/src/main/java/morawski/rafal/barcodescanner/scanner/BarcodeBoxDetector.kt
package morawski.rafal.barcodescanner.scanner

import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.util.SparseArray
import com.google.android.gms.vision.Detector
import com.google.android.gms.vision.Frame
import com.google.android.gms.vision.barcode.Barcode
import java.io.ByteArrayOutputStream

class BarcodeBoxDetector(
        private val mDelegate: Detector<Barcode>,
        private val mScannerBoxOverlay: ScannerBoxOverlay? = null
) : Detector<Barcode>() {

    override fun detect(frame: Frame): SparseArray<Barcode> {
        if (mScannerBoxOverlay == null) {
            // There is no box, so nothing to do because we don't need to crop
            // pass work to delegate
            return mDelegate.detect(frame)
        }

        val frameWidth = frame.metadata.width
        val frameHeight = frame.metadata.height
        val frameRotation = frame.metadata.rotation

        val viewWidth = mScannerBoxOverlay.width
        val viewHeight = mScannerBoxOverlay.height

        val widthScaleFactor: Float
        val heightScaleFactor: Float
        if (isPortrait) {
            widthScaleFactor = frameHeight.toFloat() / viewWidth.toFloat()
            heightScaleFactor = frameWidth.toFloat() / viewHeight.toFloat()
        } else {
            widthScaleFactor = frameWidth.toFloat() / viewWidth.toFloat()
            heightScaleFactor = frameHeight.toFloat() / viewHeight.toFloat()
        }

        val scanBoxSize = mScannerBoxOverlay.getBoxSizePx()
        val boxWidth = (scanBoxSize.width * widthScaleFactor).toInt()
        val boxHeight = (scanBoxSize.height * heightScaleFactor).toInt()

        val cropRect = getRectangle(frameWidth, frameHeight, boxWidth, boxHeight)

        val yuvImage = YuvImage(frame.grayscaleImageData?.array(), ImageFormat.NV21, frameWidth, frameHeight, null)
        val outputStream = ByteArrayOutputStream()
        yuvImage.compressToJpeg(cropRect, 100, outputStream)
        val jpegArray = outputStream.toByteArray()
        val bitmap = BitmapFactory.decodeByteArray(jpegArray, 0, jpegArray.size)

        val croppedFrame = Frame.Builder()
                .setBitmap(bitmap)
                .setRotation(frameRotation)
                .build()

        return mDelegate.detect(croppedFrame)
    }

    private fun getRectangle(frameWidth: Int, frameHeight: Int, boxWidth: Int, boxHeight: Int): Rect {
        val left: Int
        val top: Int
        val right: Int
        val bottom: Int

        if (isPortrait) {
            left = (frameWidth / 2.0 - boxHeight / 2.0).toInt()
            top = (frameHeight / 2.0 - boxWidth / 2.0).toInt()
            right = (frameWidth / 2.0 + boxHeight / 2.0).toInt()
            bottom = (frameHeight / 2.0 + boxWidth / 2.0).toInt()
        } else {
            left = (frameWidth / 2.0 - boxWidth / 2.0).toInt()
            top = (frameHeight / 2.0 - boxHeight / 2.0).toInt()
            right = (frameWidth / 2.0 + boxWidth / 2.0).toInt()
            bottom = (frameHeight / 2.0 + boxHeight / 2.0).toInt()
        }
        return Rect(left, top, right, bottom)
    }

    private val isPortrait
        get() = Resources.getSystem().configuration.orientation == Configuration.ORIENTATION_PORTRAIT
}